pkg/v1/tarball/write.go

```diff
@@ -29,80 +30,115 @@ import (
+	return MultiRefWrite(refToImage, w, opts...)
 }
 
 // Write is a wrapper to write a single image and tag to a tarball.
-func Write(ref name.Reference, img v1.Image, w io.Writer) error {
-	return MultiRefWrite(map[name.Reference]v1.Image{ref: img}, w)
+func Write(ref name.Reference, img v1.Image, w io.Writer, opts ...WriteOption) error {
+	return MultiRefWrite(map[name.Reference]v1.Image{ref: img}, w, opts...)
 }
 
 // MultiWrite writes the contents of each image to the provided reader, in the compressed format.
 // The contents are written in the following format:
 // One manifest.json file at the top level containing information about several images.
 // One file for each layer, named after the layer's SHA.
 // One file for the config blob, named after its SHA.
-func MultiWrite(tagToImage map[name.Tag]v1.Image, w io.Writer) error {
+func MultiWrite(tagToImage map[name.Tag]v1.Image, w io.Writer, opts ...WriteOption) error {
 	refToImage := make(map[name.Reference]v1.Image, len(tagToImage))
 	for i, d := range tagToImage {
 		refToImage[i] = d
 	}
-	return MultiRefWrite(refToImage, w)
+	return MultiRefWrite(refToImage, w, opts...)
 }
 
 // MultiRefWrite writes the contents of each image to the provided reader, in the compressed format.
 // The contents are written in the following format:
 // One manifest.json file at the top level containing information about several images.
 // One file for each layer, named after the layer's SHA.
 // One file for the config blob, named after its SHA.
-func MultiRefWrite(refToImage map[name.Reference]v1.Image, w io.Writer) error {
-	tf := tar.NewWriter(w)
-	defer tf.Close()
+func MultiRefWrite(refToImage map[name.Reference]v1.Image, w io.Writer, opts ...WriteOption) error {
+	// process options
+	o := &writeOptions{
+		updates: nil,
+	}
+	for _, option := range opts {
+		if err := option(o); err != nil {
+			return err
+		}
+	}
 
 	imageToTags := dedupRefToImage(refToImage)
 	var m Manifest
 
+	tw := w
+
+	// we only calculate the sizes and use a progressWriter if we were provided
+	// an option with a progress channel
+	if o.updates != nil {
+		var size int64
+		for img, name := range imageToTags {
+			manifest, err := img.Manifest()
+			if err != nil {
+				return fmt.Errorf("unable to get manifest for img %s: %v", name, err)
+			}
+			size += manifest.Config.Size
+			for _, l := range manifest.Layers {
+				size += l.Size
+			}
+		}
+
+		tw = progressWriter{
+			w:       w,
+			updates: o.updates,
+			size:    size,
+		}
+	}
+	tf := tar.NewWriter(tw)
+	defer tf.Close()
+
 	seenLayerDigests := make(map[string]struct{})
 	for img, tags := range imageToTags {
 		// Write the config.
 		cfgName, err := img.ConfigName()
 		if err != nil {
 			return err
 		}
+
 		cfgBlob, err := img.RawConfigFile()
 		if err != nil {
 			return err
 		}
-		if err := writeTarEntry(tf, cfgName.String(), bytes.NewReader(cfgBlob), int64(len(cfgBlob))); err != nil {
+		cfgSize := int64(len(cfgBlob))
```

Comment:
I'd undo this just to have a cleaner diff

Reply:
No problem.